### PR TITLE
Password guard sign up

### PR DESCRIPTION
**Pull request by dct-lau17**
-Added validation checks so:
a) all fields are filled in the sign up form
b) password must be between 6-10 characters
c) email must be in the correct format
d) user cannot enter an email address already in the database

-introduced some code in the en.yml file customise validation messages
-changed the route to be the sign up form
-updated view to show messages from the yaml file to the user

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -6,8 +6,13 @@ def new
 
   def create
     @user = User.create(user_params)
-    session[:current_user_id] = @user.id
-    redirect_to user_path(@user)
+    if @user.save
+      session[:current_user_id] = @user.id
+      redirect_to user_path(@user)
+    else
+      # flash[:alert] = "The password needs to be between 6-10 characters"
```
Comment (elishkaflint) [inline app/controllers/users_controller.rb]: can we take out this comment :)

File: app/models/user.rb [hunk 2]
```diff
@@ -1,5 +1,20 @@
+require 'uri'
 class User < ApplicationRecord
   # This is a module, part of ActiveModel library
   # full details here: https://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html
   has_secure_password
+
+  validates_length_of :password, :in => 6..10, :on => :create
+
+  validates_presence_of :first_name
```
Comment (elishkaflint) [inline app/models/user.rb]: can we DRY this out, if possible

File: app/models/user.rb [hunk 3]
```diff
@@ -1,5 +1,20 @@
...
+  validates_presence_of :last_name
+  validates_presence_of :email
+  validates_uniqueness_of :email
+
+  EmailRegex = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (elishkaflint) [inline app/models/user.rb]: is this naming convention correct? if just ordinary variable -> lower case

File: config/locales/en.yml [hunk 4]
```diff
@@ -29,5 +29,26 @@
...
+              taken: "This email address has already been used"
+              invalid: "Please provide a valid email address"
+            first_name:
+              blank: "Please enter your first name"
+            last_name:
+              blank: "Please enter your last name"
+            password:
+              too_short: "The password needs to be between 6-10 characters"
+              too_long: "The password needs to be between 6-10 characters"
+              blank: "Password must exists"
```
Comment (elishkaflint) [inline config/locales/en.yml]: must exist (typo)

File: config/locales/en.yml [hunk 5]
```diff
@@ -29,5 +29,26 @@
...
 en:
-  hello: "Hello world"
+  activerecord:
+    errors:
+      format: "%{message}"
+      models:
+        user:
+          attributes:
+            email:
+              blank: "Please enter an email address"
```
Comment (elishkaflint) [inline config/locales/en.yml]: low priority: in our agreed effort to make a bona fide facebook clone, shall we make sure the error messages reflect what facebook actually does as closely as possible?

File: spec/features/user_can_sign_in_spec.rb [hunk 6]
```diff
@@ -7,28 +7,30 @@
...
-
-    visit "/"
-    fill_in :email, with: "test_email@keepingitrails.com"
-    fill_in :password, with: "incorrect password"
-    click_button "Log In"
-    expect(page).to raise_error("Sorry, please check your password and try again")
-  end
+  # scenario "A user signs in with invalid email address" do
+  #
+  #   visit "/session/new"
```
Comment (elishkaflint) [inline spec/features/user_can_sign_in_spec.rb]: shouldn't the feature tests all start with visit '/' ?
this comment applies to a few others

File: spec/features/user_can_sign_in_spec.rb [hunk 7]
```diff
@@ -7,28 +7,30 @@
...
+  #   fill_in :email, with: "test_email_other@gmail.com"
+  #   fill_in :password, with: "pa55w0rd"
+  #   click_button "Log In"
+  #   expect(page).to have_content("Sorry, we do not recognise this email address")
+  # end
+  #
+  # scenario "A user signs in with incorrect password" do
+  #
+  #   visit "/session/new"
+  #   fill_in :email, with: "test_email@keepingitrails.com"
```
Comment (elishkaflint) [inline spec/features/user_can_sign_in_spec.rb]: you knew it was coming...... can we extract this to a web helper with very clear scenario names? we should have one web helper method for each incorrect user input (eg. email address already exists, password too long etc etc)

File: spec/features/user_can_sign_up_spec.rb [hunk 8]
```diff
@@ -7,4 +7,66 @@
...
 
+    scenario "A user provides an password with less than 6 characters" do
+      unsuccessful_sign_up_with_wrong_password
+      fill_in 'Password', with: "pass"
+      click_button "Sign Up"
+      expect(page).to have_content("The password needs to be between 6-10 characters")
+    end
+
+    scenario "A user provides an password with more than 10 characters" do
+      unsuccessful_sign_up_with_wrong_password
```
Comment (elishkaflint) [inline spec/features/user_can_sign_up_spec.rb]: these seem really thorough 👍

**Review by elishkaflint — COMMENTED** [review]
Great work daniel, super interesting implementation of error messages, have definitely learnt something new from this review!

File: app/models/user.rb (hunk 3 above)
Comment (jeff1108) [reply app/models/user.rb]: That's the naming convention

**Comment by jeff1108** [comment]
Refactor the sign up feature test with web helper
Add a test the link to the sign up page in sign in feature test
Fix the conflict between two branch
Refactor the validates of email
Add flash in the show html when the user successfully sign in and sign up  
Add two notices in the en.yml file

File: app/controllers/session_controller.rb [hunk 9]
```diff
@@ -10,13 +10,33 @@ def new
 
   #POST /session
   def create
-    @user = User.find_by(email: params[:email])
-    if @user && @user.authenticate(params[:password])
+    if valid_email? && valid_password?
       session[:current_user_id] = @user.id
-      # p @user.id
-      redirect_to user_path(@user)
+      redirect_to user_path(@user), notice: t(".notice")
```
Comment (elishkaflint) [inline app/controllers/session_controller.rb:15]: cool refactor

File: app/views/session/new.html.erb [hunk 10]
```diff
@@ -3,6 +3,8 @@
   <%= email_field_tag :email, params[:email] %>
   <%= label_tag :password, "Password" %>
   <%= password_field_tag :password, params[:password] %>
-
-<%= submit_tag "Log In" %>
+  <%= submit_tag "Log In" %>
 <% end %>
+
+<%= alert ? alert : '' %><br>
+<%= link_to "Sign up for Keeping it rails", "/users/new" %>
```
Comment (elishkaflint) [inline app/views/session/new.html.erb]: Should it be sign up for Acebook?

File: app/views/users/new.html.erb [hunk 11]
```diff
@@ -6,6 +6,14 @@
   <%= form.label :email %>
   <%= form.text_field :email %>
   <%= form.label :password %>
-  <%= form.text_field :password %>
+  <%= form.password_field :password %>
   <%= form.submit "Sign Up" %>
 <% end %>
+
+<%= link_to "Already a User?", "/session/new" %>
```
Comment (elishkaflint) [inline app/views/users/new.html.erb]: Can we change this to "Already have an Acebook account?"

**Review by elishkaflint — COMMENTED** [review]
Almost there, just a couple of tiny things, then let's MERGE 😺

**Review by elishkaflint — APPROVED** [review]
:shipit:

**Pull request merged**